feat(build.zig): add option to use system dependencies

dchinmay2 opened the pull request:
Problem:
build.zig always downloads dependencies and statically links them, which is frowned upon by distro packagers.

Solution:
Add option to use system libraries.

dchinmay2 committed:
feat(build.zig): add option to use system dependencies


Problem:
build.zig always downloads dependencies and statically links them,
which is frowned upon by distro packagers.

Solution:
Add option to use system libraries.

dchinmay2 committed:
fix(editorconfig): use 4 spaces indentation for zig files

bfredl commented on build.zig:
```diff
@@ -307,17 +307,32 @@ pub fn build(b: *std.Build) !void {
-    nvim_exe.linkLibrary(libluv);
+    if (b.systemIntegrationOption("lua", .{})) {
+        if (use_luajit) {
+            nvim_exe.root_module.linkSystemLibrary("luajit", .{});
```
```
const lualib = if (use_luajit) "luajit" else "lua5.1";
nvim_exe.root_module.linkSystemLibrary(lualib, .{});
```

p00f commented on build.zig:
```diff
@@ -307,17 +307,32 @@ pub fn build(b: *std.Build) !void {
             .optimize = optimize,
+            // system luajit fails to link otherwise, and both lua and luajit in ziglua link libc anyway:
+            // https://github.com/natecraddock/ziglua/blob/315b7f25efd29d9e2143d6c0ab685bb7f640450f/build/lua.zig#L106
+            // https://github.com/natecraddock/ziglua/blob/315b7f25efd29d9e2143d6c0ab685bb7f640450f/build/luajit.zig#L216
```
done

bfredl commented on .editorconfig:
```diff
@@ -7,6 +7,9 @@ tab_width = 8
 insert_final_newline = true
 
+[*.zig]
+indent_size = 4
```
so far I've just used `zig fmt` , and it only supports 4 spaces. There is no check in CI yet but we could add one.

p00f commented:
I tried to see if it builds with my changes on a musl distro but it looks like even the current build.zig doesn't work

dchinmay2 reviewed (commented):
re: the `install_subdir` for docs
I just got
```
./test/functional/testnvim.lua:141: Vim(help):E429: File "/home/chinmay/code/codebases/neovim/zig-out/runtime/doc/helphelp.txt" does not exist
```
when i ran `zig build functionaltest`, was the previous `install_subdir` intentional?

dchinmay2 committed:
fix(build.zig): format build.zig to keep lines under 100 chars

p00f commented:
For Arch Linux you can try this PKGBUILD
```bash
# shellcheck shell=bash
# shellcheck disable=SC2034

declare srcdir pkgdir
pkgname=neovim-zig-git
_nvim_version=0.12.0
pkgver=0.12.0.r2609.gb8221f970d
pkgrel=1
pkgdesc='Fork of Vim aiming to improve user experience, plugins, and GUIs - built using zig'
arch=(i686 x86_64 armv7h armv6h aarch64)
url='https://neovim.io'
backup=('etc/xdg/nvim/sysinit.vim')
license=('custom:neovim')
depends=(
    libluv
    libutf8proc
    libuv
    lua51-lpeg
    luajit
    tree-sitter
    tree-sitter-c
    tree-sitter-lua
    tree-sitter-markdown
    tree-sitter-query
    tree-sitter-vim
    tree-sitter-vimdoc
    unibilium
)
makedepends=('git' 'zig>=0.15.2' 'tar')
optdepends=(
    'python-pynvim: for Python plugin support (see :help python)'
    'tree-sitter-bash: tree-sitter parser for bash'
    'tree-sitter-python: tree-sitter parser for python'
    'xclip: for clipboard support on X11 (or xsel) (see :help clipboard)'
    'xsel: for clipboard support on X11 (or xclip) (see :help clipboard)'
    'wl-clipboard: for clipboard support on wayland (see :help clipboard)'
)
provides=("neovim=${_nvim_version}" 'vim-plugin-runtime')
conflicts=('neovim' 'neovim-git')
source=(
    "git+https://github.com/p00f/neovim#branch=push-nwlkmnvmmlrt"
    "ziglua.tar.gz::https://github.com/p00f/ziglua/archive/1d276472c68c96f781aa2d0189a87c46355a8049.tar.gz"
    "https://github.com/neovim/deps/raw/06ef2b58b0876f8de1a3f5a710473dcd7afff251/opt/lua-dev-deps.tar.gz"
)
noextract=("ziglua.tar.gz" "lua-dev-deps.tar.gz")
b2sums=(
    'SKIP'
    'SKIP'
    'b27aa3bb208cbb68bac8b7722fb48fc76c4b862d19d1bc7564596316a25623d727c4d8d2e520c0abe416dc78b33c6aacd4b28968206b356e4fe80691886ca48b'
    'd31cf81659e238fada8092755eb9be16f77c00a466107eb5770c6c9c32e043c91e6efada7ddb51663716a0e38ffa6e3d0093b3e6833aa961d845c7451a95491e'
    '26588b9da6459393076723bdfb8d2b16fed882070f2326bf7c35cd272dee9c18df603afb1ae2254cd0a59eff68189caf04828ef165d5de42c7a4222267604101'
)

pkgver() {
    local nvim_version_git
    cd "${srcdir}/neovim" || exit 1
    _nvim_version="$(awk -F'"' '/\.version = "/ {print $2}' build.zig.zon)"
    nvim_version_git="$(git describe --always --dirty --match 'v*.*.*' | sed -E 's/^v[0-9]+.[0-9]+.[0-9]+-//; s/^([0-9]+)-([a-z0-9]+)/\1\.\2/')"
    printf "%s.r%s\n" "$_nvim_version" "$nvim_version_git"
}

prepare() {
    local zlua_hash lua_dev_deps_hash
    zlua_hash="zlua-0.1.0-hGRpC1dCBQDf-IqqUifYvyr8B9-4FlYXqY8cl7HIetrC"
    lua_dev_deps_hash="N-V-__8AAGevEQCHAkCozca5AIdN9DFc3Luf3g3r2AcbyOrm"
    cd "${srcdir}/neovim" || exit 1
    mkdir -p zig_deps && cd zig_deps || exit 1
    mkdir -p $zlua_hash
    tar xf "${srcdir}/ziglua.tar.gz" -C $zlua_hash --strip-components=1
    mkdir -p $lua_dev_deps_hash
    tar xf "${srcdir}/lua-dev-deps.tar.gz" -C $lua_dev_deps_hash --warning=no-unknown-keyword
}

build() {
    cd "${srcdir}/neovim" || exit 1
    zig build --system zig_deps -Doptimize=ReleaseFast -Dcpu=native
}

check() {
    cd "${srcdir}/neovim" || exit 1
    zig-out/bin/nvim --version
    zig-out/bin/nvim --headless -u NONE -i NONE -c ':quit'
}

package() {
    pushd . >/dev/null
    cd "${srcdir}/neovim" || exit 1
    zig build install --system zig_deps --prefix "${pkgdir}/usr" -Doptimize=ReleaseFast -Dcpu=native
    rm "${pkgdir}/usr/bin/nlua0"
    rm -r "${pkgdir}/usr/runtime"

    install -Dm644 LICENSE.txt -t "${pkgdir}/usr/share/licenses/${pkgname}/"
    install -Dm644 runtime/nvim.desktop -t "${pkgdir}/usr/share/applications/"
    install -Dm644 runtime/nvim.appdata.xml -t "${pkgdir}/usr/share/metainfo/"
    install -Dm644 runtime/nvim.png -t "${pkgdir}/usr/share/pixmaps/"
    # shellcheck disable=SC2164
    popd >/dev/null

    # Make Arch Vim packages work
    mkdir -p "${pkgdir}"/etc/xdg/nvim
    echo "\" This line makes pacman-installed global Arch Linux vim packages work." >"${pkgdir}"/etc/xdg/nvim/sysinit.vim
    echo "source /usr/share/nvim/archlinux.vim" >>"${pkgdir}"/etc/xdg/nvim/sysinit.vim

    mkdir -p "${pkgdir}"/usr/share/vim
    echo "set runtimepath+=/usr/share/vim/vimfiles" >"${pkgdir}"/usr/share/nvim/archlinux.vim
}

# vim:set sw=4 sts=4 et:

```

p00f commented:
At least on my Arch Linux system, unittests all pass with the system integration build. If I disable the system integration for lua, vimdoc, and query parsers (because the distro doesn't have the versions used by neovim) using `-fno-sys=tree-sitter-lua -fno-sys=tree-sitter-vimdoc -fno-sys=tree-sitter-query`, as many functionaltests pass as with the default zig build (all system integrations disabled)

These are the failures
```
FAILED   7 tests, listed below:
FAILED   test/functional/api/vim_spec.lua @ 3401: API nvim_get_runtime_file can find files
./test/functional/api/vim_spec.lua:3423: Expected objects to be the same.
Passed in:
(number) 2
Expected:
(number) 1

stack traceback:
        ./test/functional/api/vim_spec.lua:3423: in function <./test/functional/api/vim_spec.lua:3401>

FAILED   test/functional/api/vim_spec.lua @ 3431: API nvim_get_runtime_file can find directories
./test/functional/api/vim_spec.lua:3433: Expected objects to be the same.
Passed in:
(number) 2
Expected:
(number) 1

stack traceback:
        ./test/functional/api/vim_spec.lua:3433: in function <./test/functional/api/vim_spec.lua:3431>

FAILED   test/functional/editor/K_spec.lua @ 46: K <esc> kills the buffer for a running 'keywordprg' command
./test/functional/editor/K_spec.lua:56: Expected objects to be the same.
Passed in:
(string) 'n'
Expected:
(string) 't'

stack traceback:
        ./test/functional/editor/K_spec.lua:56: in function <./test/functional/editor/K_spec.lua:46>

FAILED   test/functional/ex_cmds/mksession_spec.lua @ 197: :mksession restores CWD for :terminal buffer at root directory #16988
./test/functional/ex_cmds/mksession_spec.lua:216: Row 1 did not match.
Expected:
  |*^/                                                 |
  |*                                                  |
  |*[Process exited 0]                                |
  |*                                                  |
  |*                                                  |
  |                                                  |
Actual:
  |*^error: can not save history                       |
  |*warning-path: Unable to locate data directory deri|
  |*ved from $XDG_DATA_HOME: 'Xtest_xdg/share/fish'.  |
  |*warning-path: The error was 'Permission denied'.  |
  |*warning-path: Please set $XDG_DATA_HOME to a direc|
  |                                                  |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

Snapshot:
screen:expect([[
  ^error: can not save history                       |
  warning-path: Unable to locate data directory deri|
  ved from $XDG_DATA_HOME: 'Xtest_xdg/share/fish'.  |
  warning-path: The error was 'Permission denied'.  |
  warning-path: Please set $XDG_DATA_HOME to a direc|
                                                    |
]])

stack traceback:
        ./test/functional/ui/screen.lua:906: in function '_wait'
        ./test/functional/ui/screen.lua:536: in function 'expect'
        ./test/functional/ex_cmds/mksession_spec.lua:216: in function <./test/functional/ex_cmds/mksession_spec.lua:197>

FAILED   test/functional/terminal/tui_spec.lua @ 1602: TUI paste: terminal mode
./test/functional/terminal/tui_spec.lua:1609: Row 1 did not match.
Expected:
  |*tty ready                                         |
  |*^                                                  |
  |*                                                  |
  |                                                  |
  |{109:^^^^^^^                                           }|
  |{5:-- TERMINAL --}                                    |
  |{5:-- TERMINAL --}                                    |
Actual:
  |*process does not own the terminal                 |
  |*                                                  |
  |*[Process exited 2]^                                |
  |                                                  |
  |{109:^^^^^^^                                           }|
  |{5:-- TERMINAL --}                                    |
  |{5:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

Snapshot:
screen:expect([[
  process does not own the terminal                 |
                                                    |
  [Process exited 2]^                                |
                                                    |
  {109:^^^^^^^                                           }|
  {5:-- TERMINAL --}                                    |*2
]])

stack traceback:
        ./test/functional/ui/screen.lua:906: in function '_wait'
        ./test/functional/ui/screen.lua:536: in function 'expect'
        ./test/functional/terminal/tui_spec.lua:1609: in function <./test/functional/terminal/tui_spec.lua:1602>

FAILED   test/functional/terminal/tui_spec.lua @ 2107: TUI forwards :term palette colors with termguicolors
./test/functional/terminal/tui_spec.lua:2145: Row 1 did not match.
Expected:
  |*^tty ready                                         |
  |                                                  |
  |*                                                  |
  |                                                  |
  |{2:^^^^^^^                                           }|
  |                                                  |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*^process does not own the terminal                 |
  |                                                  |
  |*[Process exited 2]                                |
  |                                                  |
  |{2:^^^^^^^                                           }|
  |                                                  |
  |{3:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

Snapshot:
screen:expect([[
  ^process does not own the terminal                 |
                                                    |
  [Process exited 2]                                |
                                                    |
  {2:^^^^^^^                                           }|
                                                    |
  {3:-- TERMINAL --}                                    |
]])

stack traceback:
        ./test/functional/ui/screen.lua:906: in function '_wait'
        ./test/functional/ui/screen.lua:536: in function 'expect'
        ./test/functional/terminal/tui_spec.lua:2145: in function <./test/functional/terminal/tui_spec.lua:2107>

FAILED   test/functional/vimscript/has_spec.lua @ 67: has() "terminfo"
./test/functional/vimscript/has_spec.lua:75: Expected objects to be the same.
Passed in:
(number) 1
Expected:
(number) 0

stack traceback:
        ./test/functional/vimscript/has_spec.lua:75: in function <./test/functional/vimscript/has_spec.lua:67>
```

p00f commented:
re: the failing functionaltests `zig build functionaltest -- test/functional/ex_cmds/make_spec.lua` passes locally

p00f commented on build.zig:
```diff
@@ -280,51 +382,134 @@ pub fn build(b: *std.Build) !void {
+    try unittest_include_path.append(b.allocator, gen_config.getDirectory());
+    if (system_integration_options.lua) {
+        if (use_luajit) {
+            try appendSystemIncludePath(b, &unittest_include_path, "luajit");
```
done

dchinmay2 committed:
feat(build.zig): install parsers in install step

Problem: The install step does not install parsers
Solution: Make the install step depend on installing parsers

bfredl commented on build.zig:
```diff
@@ -435,27 +625,51 @@ pub fn build(b: *std.Build) !void {
+    test_deps.dependOn(add_ts_parser(b, "markdown", parser_markdown.path("tree-sitter-markdown/"), true, target, optimize, .test_));
+    install.dependOn(add_ts_parser(b, "markdown", parser_markdown.path("tree-sitter-markdown/"), true, target, optimize, .install));
+    test_deps.dependOn(add_ts_parser(b, "markdown_inline", parser_markdown.path("tree-sitter-markdown-inline/"), true, target, optimize, .test_));
+    install.dependOn(add_ts_parser(b, "markdown", parser_markdown.path("tree-sitter-markdown-inline/"), true, target, optimize, .install));
```
should this be `markdown_inline` ?

dchinmay2 committed:
docs(build.zig): document zig build

p00f commented:
@justinmk done
